### PR TITLE
ENH add weights for Group Lasso

### LINKED ISSUE #219: Adaptive GroupLasso
Hi, 
thanks for this great package! Does the group-lasso method allow for different penalizations for different groups, such that one could fit an adaptive group lasso, i.e., applying different weights in the second call to the method? (reference for adapative group lasso: https://www.sciencedirect.com/science/article/abs/pii/S0167947308002582)

If not, would it be possible to extend the method (easily)?

Thank you.

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
closes #219

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: celer-env
+dependencies:
+  - python>=3.8
+  - pip
+  # dependencies
+  - Cython>=0.26
+  # - download
```
Comment (mathurinm) [inline environment.yml:7]: those must be installed with pip. It's doable in this file, see https://stackoverflow.com/questions/35245401/combining-conda-environment-yml-with-pip-requirements-txt

File: celer/dropin_sklearn.py [hunk 2]
```diff
@@ -752,6 +752,10 @@ class GroupLasso(Lasso_sklearn):
     fit_intercept : bool, optional (default=True)
         Whether or not to fit an intercept.
 
+    weights : array, shape (n_features,), optional (default=None)
```
Comment (mathurinm) [inline celer/dropin_sklearn.py:755]: It's rather of shape (n_groups,)

File: celer/dropin_sklearn.py [hunk 3]
```diff
@@ -752,6 +752,10 @@ class GroupLasso(Lasso_sklearn):
...
+        Strictly positive weights used in the L1 penalty part of the Lasso
```
Comment (mathurinm) [inline celer/dropin_sklearn.py:756]: not a Lasso, not a L1 penalty

File: celer/group_fast.pyx [hunk 4]
```diff
@@ -25,29 +25,34 @@ cdef:
...
     for g in range(n_groups):
         nrm = 0.
+
         for k in range(grp_ptr[g], grp_ptr[g + 1]):
             j = grp_indices[k]
             nrm += w[j] ** 2
-        p_obj += alpha * sqrt(nrm)
+        p_obj += sqrt(nrm) * weights[g]
+
+    p_obj *= alpha
```
Comment (mathurinm) [inline celer/group_fast.pyx:43]: This induces an error, when doing this you also multiply the datafit by alpha

File: celer/group_fast.pyx [hunk 5]
```diff
@@ -25,29 +25,34 @@ cdef:
...
     return p_obj
 
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
+# TODO add weights (done)
```
Comment (mathurinm) [inline celer/group_fast.pyx:50]: rm these comments

File: celer/group_fast.pyx [hunk 6]
```diff
@@ -63,6 +68,10 @@ cpdef floating dnorm_grp(
 
     if ws_size == n_groups:  # max over all groups
         for g in range(n_groups):
+            # skip infinit weights
```
Comment (mathurinm) [inline celer/group_fast.pyx:71]: typo

File: celer/group_fast.pyx [hunk 7]
```diff
@@ -79,10 +88,14 @@ cpdef floating dnorm_grp(
                                     &inc)
                 tmp += Xj_theta ** 2
 
-            scal = max(scal, sqrt(tmp))
+            scal = max(scal, sqrt(tmp) / weights[g])
 
     else:  # scaling only with features in C
         for g_idx in range(ws_size):
+            # skip infinit weights
```
Comment (mathurinm) [inline celer/group_fast.pyx:95]: typo

File: celer/group_fast.pyx [hunk 8]
```diff
@@ -365,7 +383,8 @@ cpdef celer_grp(
 
                 if d_obj_in > highest_d_obj_in:
                     highest_d_obj_in = d_obj_in
-                p_obj_in = primal_grplasso(alpha, R, grp_ptr, grp_indices, w)
+                # TODO: potontially add weights (done)
```
Comment (mathurinm) [inline celer/group_fast.pyx:386]: rm

File: celer/homotopy.py [hunk 9]
```diff
@@ -144,6 +144,7 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
         (Is returned only when ``return_thetas`` is set to True).
     """
 
+    # type of the problem
```
Comment (mathurinm) [inline celer/homotopy.py:147]: rm

File: celer/homotopy.py [hunk 10]
```diff
@@ -182,6 +183,7 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
 
     X_dense, X_data, X_indices, X_indptr = _sparse_and_dense(X)
 
+    # set weights
```
Comment (mathurinm) [inline celer/homotopy.py:186]: rm

File: celer/homotopy.py [hunk 11]
```diff
@@ -192,6 +194,7 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
             f"Expected {X.shape[1]}, got {weights.shape[0]}."
         )
 
+    # set alphas of the path
```
Comment (mathurinm) [inline celer/homotopy.py:197]: rm

File: celer/homotopy.py [hunk 12]
```diff
@@ -201,6 +204,7 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
                                  positive) / 2
         elif pb == GRPLASSO:
             # TODO compute it with dscal to handle centering sparse
+            # TODO consider weights when computing alpha_max
```
Comment (mathurinm) [inline celer/homotopy.py:207]: done?

File: environment.yml [hunk 13]
```diff
@@ -0,0 +1,20 @@
+name: celer-env
```
Comment (mathurinm) [inline environment.yml:1]: rm this file

File: tt_numpydoc.py [hunk 14]
```diff
@@ -0,0 +1 @@
+from numpydoc import docscrape
```
Comment (mathurinm) [inline tt_numpydoc.py:1]: rm

**Review by mathurinm — CHANGES_REQUESTED** [review]

File: celer/group_fast.pyx (hunk 4 above)
Comment (Badr-MOUFAD) [reply celer/group_fast.pyx:43]: Fixed!
Yet, I am still a bit confused about that since ``a*(b+c+d) = a*b+a*c+a*d`` (basic math).
Also, I believe that the left-hand side way of computing (3 operations) is more effective than the right-hand side (5 operations). More precisely if we have a sum with ``n`` terms, the left-hand side needs ``n`` operation to be computed whereas the right-hand side needs ``n + n - 1``.

Comment (mathurinm) [reply celer/group_fast.pyx:43]: 1) this is not a bottleneck of the computation O(p) vs O(np) for the critical part.
2) as I said, with your method the datafit is multiplied by alpha. It should not.

File: celer/homotopy.py (hunk 12 above)
Comment (Badr-MOUFAD) [reply celer/homotopy.py:207]: not yet, 
if I m not mistaken, I should divide by weight

```python
elif pb == GRPLASSO:
            # TODO compute it with dscal to handle centering sparse
            # TODO consider weights when computing alpha_max
            alpha_max = 0
            for g in range(n_groups):
                X_g = X[:, grp_indices[grp_ptr[g]:grp_ptr[g + 1]]]
                alpha_max = max(alpha_max, norm(X_g.T @ y / weights[g], ord=2))
            alpha_max /= n_samples
```

Comment (mathurinm) [reply celer/homotopy.py:207]: yes

File: celer/group_fast.pyx [hunk 15]
```diff
@@ -271,6 +281,7 @@ cpdef celer_grp(
         # elif pb == LOGREG:
             # radius = sqrt(gap / 2.) / alpha
 
+        # TODO potontially add weights
```
Comment (Badr-MOUFAD) [inline celer/group_fast.pyx:284]: Do we need to implement weights for ``set_prios_grp``?

Comment (mathurinm) [reply celer/group_fast.pyx:284]: yes, as we do for set_prios

File: celer/group_fast.pyx [hunk 16]
```diff
@@ -254,7 +264,7 @@ cpdef celer_grp(
             highest_d_obj = d_obj
             # TODO implement a best_theta
 
-        p_obj = primal_grplasso(alpha, R, grp_ptr, grp_indices, w)
+        p_obj = primal_grplasso(alpha, R, grp_ptr, grp_indices, w, weights)
```
Comment (Badr-MOUFAD) [inline celer/group_fast.pyx:267]: Up

File: celer/group_fast.pyx [hunk 17]
```diff
@@ -63,6 +66,10 @@ cpdef floating dnorm_grp(
 
     if ws_size == n_groups:  # max over all groups
         for g in range(n_groups):
+            # skip infinite weights
```
Comment (mathurinm) [inline celer/group_fast.pyx:69]: I think the following lines are simple enough to be understood without comment

File: celer/group_fast.pyx [hunk 18]
```diff
@@ -79,10 +86,14 @@ cpdef floating dnorm_grp(
                                     &inc)
                 tmp += Xj_theta ** 2
 
-            scal = max(scal, sqrt(tmp))
+            scal = max(scal, sqrt(tmp) / weights[g])
 
     else:  # scaling only with features in C
         for g_idx in range(ws_size):
+            # skip infinite weights
```
Comment (mathurinm) [inline celer/group_fast.pyx:93]: same

File: celer/tests/test_mtl.py [hunk 19]
```diff
@@ -57,6 +57,7 @@ def test_GroupLasso_MultitaskLasso_equivalence():
     X_data = np.empty([1], dtype=X.dtype)
     X_indices = np.empty([1], dtype=np.int32)
     X_indptr = np.empty([1], dtype=np.int32)
+    # TODO add weights
```
Comment (Badr-MOUFAD) [inline celer/tests/test_mtl.py:60]: FYI: All tests are passed except for this one. 
I need to pass `` weights`` argument to ``dnorm_grp``

Comment (mathurinm) [reply celer/tests/test_mtl.py:60]: Great ! When this is done, you'll need to add a test to check that weights behave as expected for GroupLasso. See how it's done for the Lasso here: https://github.com/mathurinm/celer/blob/main/celer/tests/test_lasso.py#L187

File: celer/tests/test_docstring_parameters.py [hunk 20]
```diff
@@ -8,6 +8,7 @@
 from pkgutil import walk_packages
 from inspect import getsource
 
+from numpydoc import docscrape
```
Comment (Badr-MOUFAD) [inline celer/tests/test_docstring_parameters.py:11]: why not import ``numpydoc`` globally?

File: celer/tests/test_docstring_parameters.py [hunk 21]
```diff
@@ -108,7 +109,7 @@ def check_parameters_match(func, doc=None):
 # @requires_numpydoc
 def test_docstring_parameters():
```
Comment (Badr-MOUFAD) [inline celer/tests/test_docstring_parameters.py:110]: I cannot pass this test either locally or in the test workflow.
In both cases, I get the following error

```bash
celer\tests\test_docstring_parameters.py:111: ModuleNotFoundError
```

File: celer/tests/test_mtl.py [hunk 22]
```diff
@@ -199,5 +202,52 @@ def test_GroupLassoCV(sparse_X):
     check_estimator(clf)
 
 
+def test_weights_group_lasso():
```
Comment (Badr-MOUFAD) [inline celer/tests/test_mtl.py:205]: The test takes a lot of time to run and always fails.

File: celer/tests/test_docstring_parameters.py (hunk 21 above)
Comment (mathurinm) [reply celer/tests/test_docstring_parameters.py:110]: do you have numpydoc installed ?

Comment (mathurinm) [reply celer/tests/test_docstring_parameters.py:110]: On the CI, it fails because some docstring parameters are not in the same order as i the function definition: see https://github.com/mathurinm/celer/runs/5746219541?check_suite_focus=true#step:5:208

File: celer/tests/test_mtl.py [hunk 23]
```diff
@@ -57,9 +58,11 @@ def test_GroupLasso_MultitaskLasso_equivalence():
     X_data = np.empty([1], dtype=X.dtype)
     X_indices = np.empty([1], dtype=np.int32)
     X_indptr = np.empty([1], dtype=np.int32)
+    weights = np.ones(len(grp_ptr))
```
Comment (mathurinm) [inline celer/tests/test_mtl.py:61]: this has 1 too many elements

File: celer/tests/test_mtl.py [hunk 24]
```diff
@@ -199,5 +202,52 @@ def test_GroupLassoCV(sparse_X):
     check_estimator(clf)
 
 
+def test_weights_group_lasso():
+    # generate data
+    sparse_X = 1
```
Comment (mathurinm) [inline celer/tests/test_mtl.py:207]: pass directly as parameter and use True. Better yet, parametrize the test with pytest.mark.parametrize to use both sparse_X=Frue and False

File: celer/tests/test_mtl.py [hunk 25]
```diff
@@ -199,5 +202,52 @@ def test_GroupLassoCV(sparse_X):
...
+
+    # generate weihgts
+    groups = 5
+    n_groups = n_features // groups
+    np.random.seed(0)
+    weights = np.abs(np.random.randn(n_groups))
+
+    tol = 1e-8
+    params = {'n_alphas': 10, 'tol': tol, 'verbose': 1}
+    augmented_weights = (weights[:, None] *
```
Comment (mathurinm) [inline celer/tests/test_mtl.py:219]: np.repeat is less error prone here

File: celer/tests/test_mtl.py [hunk 26]
```diff
@@ -199,5 +202,52 @@ def test_GroupLassoCV(sparse_X):
...
+                         np.ones((n_groups, groups))).reshape((-1, ))
+
+    # method 1
+    alphas1, coefs1, gaps1 = celer_path(
+        X, y, "grouplasso", groups=groups, weights=weights,
+        eps=1e-2, **params)
+    # method 2
+    alphas2, coefs2, gaps2 = celer_path(
+        X.multiply(1 / augmented_weights[None, :]), y, "grouplasso",
```
Comment (mathurinm) [inline celer/tests/test_mtl.py:228]: The mistake may be here : you pass both the weights and the scaled matrix, so you weight twice. Compare with the Lasso test

File: celer/tests/test_mtl.py [hunk 27]
```diff
@@ -199,5 +202,52 @@ def test_GroupLassoCV(sparse_X):
...
+        groups=groups, eps=1e-2, **params)
+
+    assert_allclose(alphas1, alphas2)
+    assert_allclose(
+        coefs1, coefs2 / augmented_weights[:, None], atol=1e-4, rtol=1e-3)
+    assert_array_less(gaps1, tol * norm(y) ** 2 / len(y))
+    assert_array_less(gaps2, tol * norm(y) ** 2 / len(y))
+
+    # fit
+    alpha = 0.001
```
Comment (mathurinm) [inline celer/tests/test_mtl.py:238]: don't use an absolute alpha

File: celer/tests/test_mtl.py [hunk 28]
```diff
@@ -199,5 +202,52 @@ def test_GroupLassoCV(sparse_X):
...
+                      fit_intercept=False).fit(X, y)
+    clf2 = GroupLasso(groups, alpha=alpha, fit_intercept=False).fit(
+        X.multiply(1. / augmented_weights), y)
+
+    assert_allclose(clf1.coef_, clf2.coef_ / augmented_weights)
+
+    # weights must be > 0
+    clf1.weights[0] = 0.
+    np.testing.assert_raises(ValueError, clf1.fit, X=X, y=y)
+    return
```
Comment (mathurinm) [inline celer/tests/test_mtl.py:249]: rm

File: celer/tests/test_mtl.py [hunk 29]
```diff
@@ -230,23 +228,17 @@ def test_weights_group_lasso():
...
-    clf2 = GroupLasso(groups, alpha=alpha, fit_intercept=False).fit(
-        X.multiply(1. / augmented_weights), y)
-
-    assert_allclose(clf1.coef_, clf2.coef_ / augmented_weights)
 
-    # weights must be > 0
-    clf1.weights[0] = 0.
-    np.testing.assert_raises(ValueError, clf1.fit, X=X, y=y)
-    return
+def test_check_weights():
```
Comment (Badr-MOUFAD) [inline celer/tests/test_mtl.py:236]: To test whether GroupLasso raises an error if any of the weights is null. Isn't it?

File: celer/dropin_sklearn.py [hunk 30]
```diff
@@ -752,6 +752,10 @@ class GroupLasso(Lasso_sklearn):
     fit_intercept : bool, optional (default=True)
         Whether or not to fit an intercept.
 
+    weights : array, shape (n_groups,), optional (default=None)
```
[the remark this answers is not in the document]
Comment (Badr-MOUFAD) [reply celer/dropin_sklearn.py:755]: You are right. I swapped the two arguments and now I pass the test locally.

File: celer/tests/test_mtl.py (hunk 29 above)
Comment (mathurinm) [reply celer/tests/test_mtl.py:236]: I don't get your question

File: celer/tests/test_mtl.py (hunk 26 above)
Comment (Badr-MOUFAD) [reply celer/tests/test_mtl.py:226]: I cannot see what you mean. I scale the matrix only once. 
I followed the same implementation as ``test_weights`` in ``test_lasso.py``.

File: celer/homotopy.py [hunk 31]
```diff
@@ -182,15 +183,22 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
 
     X_dense, X_data, X_indices, X_indptr = _sparse_and_dense(X)
 
-    if weights is None:
+    if weights is None and pb in (LASSO, LOGREG):
```
[the remark this answers is not in the document]
Comment (Badr-MOUFAD) [reply celer/homotopy.py:186]: I will implement also a ``_check_pb``.
I don't think it is a good idea to implement the same thing for ``alphas``. Indeed, the ``_check_alphas`` function would have a lot of arguments.

[the remark this answers is not in the document]
Comment (mathurinm) [reply celer/homotopy.py:186]: It's generally a good idea to keep the PR focussed on a single problem, here adding the weights to the group lasso. Other modifications belong to other PR.
Otherwise you end up with huge PRs which have less chances of being merged.
So no check_pb in this PR

File: celer/tests/test_mtl.py (hunk 29 above)
Comment (Badr-MOUFAD) [reply celer/tests/test_mtl.py:236]: What is the purpose of this test?

File: celer/homotopy.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (Badr-MOUFAD) [reply celer/homotopy.py:186]: Got you! 
However, I have already implemented it. Let's make an exception only this time

File: celer/tests/test_mtl.py (hunk 29 above)
Comment (mathurinm) [reply celer/tests/test_mtl.py:236]: Ah, ok ! Yes, it is the purpose of this test.
Each test should test a précisé aspect of the code, so i isolated it from the solver test

File: celer/homotopy.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (mathurinm) [reply celer/homotopy.py]: No, let's not. It does not belong to this PR. It needs it's separate discussion and I am not even sure it's a good idea. It will make reviewing and discussing longer. In fact, it already is.

**Comment by mathurinm** [comment]
The automatic build of the documentation fails. Can you check why and fix it ?

**Comment by mathurinm** [comment]
It's unrelated to this PR. #224 will fix it and I will merge main into this branch.

If we are all set can you clean up the various comments etc ?

File: celer/tests/test_mtl.py (hunk 26 above)
Comment (Badr-MOUFAD) [reply celer/tests/test_mtl.py:226]: I m still unable to fix this test
Everything seems right, yet the test always fails

**Comment by Badr-MOUFAD** [comment]
Sure, I will do it!
Could you check this https://github.com/mathurinm/celer/pull/223/files#r838150121
I m still struggling with ``test_weights_group_lasso``. It's the only reason why the build fails.

**Comment by mathurinm** [comment]
Do you see a test failure locally ?
The tests pass on GH actions : https://github.com/mathurinm/celer/runs/5758904704?check_suite_focus=true

File: celer/tests/test_mtl.py (hunk 26 above)
Comment (mathurinm) [reply celer/tests/test_mtl.py:226]: where do you see failure ? with which error msg ?

File: celer/tests/test_docstring_parameters.py [hunk 32]
```diff
@@ -108,7 +109,7 @@ def check_parameters_match(func, doc=None):
...
     """Test module docstring formatting."""
-    from numpydoc import docscrape
+    # from numpydoc import docscrape
```
Comment (mathurinm) [inline celer/tests/test_docstring_parameters.py:112]: rm the whole line

File: celer/tests/test_docstring_parameters.py [hunk 33]
```diff
@@ -65,7 +66,7 @@ def get_name(func):
 
 def check_parameters_match(func, doc=None):
     """Check docstring, return list of incorrect results."""
-    from numpydoc import docscrape
+    # from numpydoc import docscrape
```
Comment (mathurinm) [inline celer/tests/test_docstring_parameters.py:69]: rm the line

File: celer/homotopy.py [hunk 34]
```diff
@@ -325,6 +310,58 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
...
+        raise ValueError("Unsupported problem: %s" % pb)
+
+    return pb, grp_ptr, grp_indices, n_groups
+
+
+def _check_weights(weights, pb, X, n_groups):
+    """Handle weights cases."""
+    if weights is None:
+        n_weights = n_groups if pb == GRPLASSO else X.shape[1]
+
```
Comment (mathurinm) [inline celer/homotopy.py:348]: rm blank line

File: celer/homotopy.py [hunk 35]
```diff
@@ -325,6 +310,58 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
...
+        weights = np.ones(n_weights).astype(X.dtype)
+    # weights are provided
```
Comment (mathurinm) [inline celer/homotopy.py:350]: rm this comment

File: celer/homotopy.py [hunk 36]
```diff
@@ -325,6 +310,58 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
...
+    elif (weights <= 0).any():
+        raise ValueError("0 or negative weights are not supported.")
+    else:
+        expected_n_weights = n_groups if pb == GRPLASSO else X.shape[1]
+
+        if weights.shape[0] != expected_n_weights:
+            raise ValueError(
+                "As many weights as groups must be passed. "
```
Comment (mathurinm) [inline celer/homotopy.py:358]: you raise an error message with "groups" in the case of the Lasso, this will confuse the user

File: celer/homotopy.py [hunk 37]
```diff
@@ -325,6 +310,58 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
     return results
 
 
+def _check_pb(pb, X, y, groups):
```
Comment (mathurinm) [inline celer/homotopy.py:313]: revert the changes in this function. Do it in another PR if you want

File: celer/homotopy.py [hunk 38]
```diff
@@ -21,6 +21,12 @@
 LOGREG = 1
 GRPLASSO = 2
 
+DICT_PB = {
```
Comment (mathurinm) [inline celer/homotopy.py:24]: rm

File: celer/tests/test_mtl.py (hunk 26 above)
Comment (Badr-MOUFAD) [reply celer/tests/test_mtl.py:226]: The two methods with weights and scaling with weights don't give the same result.
So, I get an Assertion error as the following:

```bash
E       AssertionError: 
E       Not equal to tolerance rtol=0.001, atol=0
E       
E       Mismatch: 45%
E       Max absolute difference: 2.90401975
E       Max relative difference: 125.45373015
E        x: array([[ 0.      ,  0.      ,  0.      ,  0.      ,  0.008718,  0.189253,
E                0.355291,  0.621276,  1.078897,  1.430047],
E              [ 0.      ,  0.      ,  0.      ,  0.      ,  0.005401,  0.112003,...
E        y: array([[ 0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
E                0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
E                0.000000e+00,  0.000000e+00],...

celer\tests\test_mtl.py:231: AssertionError
```

Comment (mathurinm) [reply celer/tests/test_mtl.py:226]: Did you recompile the cython code ? Run pip install -e .

Comment (Badr-MOUFAD) [reply celer/tests/test_mtl.py:226]: My bad!
Now all is good

File: celer/tests/test_mtl.py (hunk 29 above)
Comment (Badr-MOUFAD) [reply celer/tests/test_mtl.py:236]: Does it make the test more robust if the null weight is set randomly?

File: celer/tests/test_mtl.py (hunk 23 above)
Comment (Badr-MOUFAD) [reply celer/tests/test_mtl.py:61]: should I decrease the size of weights by 1?

File: celer/homotopy.py [hunk 39]
```diff
@@ -144,16 +144,19 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
         (Is returned only when ``return_thetas`` is set to True).
     """
 
-    if pb.lower() not in ("lasso", "logreg", "grouplasso"):
+    pb = pb.lower()
+    if pb not in ("lasso", "logreg", "grouplasso"):
         raise ValueError("Unsupported problem %s" % pb)
-    if pb.lower() == "lasso":
+    # prevent referenced before assigned error
+    grp_ptr, grp_indices, n_groups = None, None, None
```
Comment (mathurinm) [inline celer/homotopy.py:151]: can you avoid diff like that ? They make the PR harder to review. Can you revert ?

File: celer/homotopy.py [hunk 40]
```diff
@@ -144,16 +144,19 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
...
+    if pb == "lasso":
         pb = LASSO
-    elif pb.lower() == "logreg":
+    elif pb == "logreg":
         pb = LOGREG
         if set(y) - set([-1.0, 1.0]):
             raise ValueError(
                 "y must contain only -1. or 1 values. Got %s " % (set(y)))
-    elif pb.lower() == "grouplasso":
+    elif pb == "grouplasso":
```
Comment (mathurinm) [inline celer/homotopy.py:159]: same here, it's globally a good idea to do it but it does not belong in this PR. Keep the diff as simple as possible.

File: celer/homotopy.py [hunk 41]
```diff
@@ -182,15 +185,8 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
...
-    if weights is None:
-        weights = np.ones(n_features).astype(X.dtype)
-    elif (weights <= 0).any():
-        raise ValueError("0 or negative weights are not supported.")
-    elif weights.shape[0] != X.shape[1]:
-        raise ValueError(
-            "As many weights as features must be passed. "
-            f"Expected {X.shape[1]}, got {weights.shape[0]}."
-        )
+    # handle weight cases
```
Comment (mathurinm) [inline celer/homotopy.py:188]: remove comments like that. the function name should be explicit enough: you check the weights.

File: celer/homotopy.py [hunk 42]
```diff
@@ -283,21 +279,22 @@ def celer_path(X, y, pb, eps=1e-3, n_alphas=100, alphas=None,
                 scal = dnorm_l1(X, theta, weights, X_sparse_scaling,
                                 positive)
             elif pb == GRPLASSO:
+                # TODO add weights to dnrom_grp (done)
```
Comment (mathurinm) [inline celer/homotopy.py:282]: remove

File: celer/homotopy.py (hunk 34 above)
Comment (mathurinm) [reply celer/homotopy.py:329]: to address

File: celer/tests/test_mtl.py (hunk 23 above)
Comment (mathurinm) [reply celer/tests/test_mtl.py:61]: yes, grp_ptr has shape (n_groups + 1,)

File: celer/tests/test_mtl.py (hunk 29 above)
Comment (mathurinm) [reply celer/tests/test_mtl.py:236]: Maybe marginally. It's not worth doing it.

File: celer/homotopy.py (hunk 39 above)
Comment (Badr-MOUFAD) [reply celer/homotopy.py:150]: ``grp_ptr, grp_indices, n_groups = None, None, None`` is necessary to avoid breaking down``_check_weights``

Comment (mathurinm) [reply celer/homotopy.py:150]: `grp_ptr, grp_indices` are not needed in `check_weights`, so we can remove these 2

Fewer variable definitions makes simpler code, fewer sources of mistake.

Can you think of a design where you would'nt have to do this ?

Comment (Badr-MOUFAD) [reply celer/homotopy.py:150]: I added these two variables for code readability.
I will remove them and keep ``n_groups``

**Review by mathurinm — APPROVED** [review]
